File: go/leet_code/623.go
// Package leet_code provides solutions for LeetCode problems.
package leet_code

// TreeNode represents a node in a binary tree.
// Each node contains an integer value, a pointer to the left child node, and a pointer to the right child node.
type TreeNode struct {
	Val   int       // The value of the node.
	Left  *TreeNode // Pointer to the left child node.
	Right *TreeNode // Pointer to the right child node.
}

// addOneRow adds a row of nodes with a given value at a given depth in the binary tree.
// The original nodes at depth will be the left child of the new nodes.
// The root of the tree will be the new nodes if depth is 1.
// The new row will be inserted immediately below the root if depth is 2.
// If depth is greater than 2, the function is called recursively for the left and right child nodes with a decremented depth.
//
// Parameters:
//  - root: Pointer to the root node of the binary tree.
//  - val: The value for the new nodes.
//  - depth: The depth at which the new row should be added.
//
// Returns:
//  - Pointer to the root node of the modified binary tree.
//
// Error Handling:
//  - If the root is nil, the function returns nil.
//
// Example Usage:
//  root := &TreeNode{Val: 1, Left: &TreeNode{Val: 2}, Right: &TreeNode{Val: 3}}
//  val := 4
//  depth := 2
//  newRoot := addOneRow(root, val, depth) // newRoot now points to the root of the modified tree.
func addOneRow(root *TreeNode, val int, depth int) *TreeNode {
	if root == nil {
		return nil
	}

	if depth == 1 {
		root = &TreeNode{Val: val, Left: root}
	} else if depth == 2 {
		root.Left = &TreeNode{Val: val, Left: root.Left}
		root.Right = &TreeNode{Val: val, Right: root.Right}
	} else {
		addOneRow(root.Left, val, depth-1)
		addOneRow(root.Right, val, depth-1)
	}
	return root
}

File: swift/leet_code/623.swift
// MARK: - TreeNode
/**
 A class that represents a node in a binary tree.
 Each node has a value and two children: left and right.
 */
public class TreeNode {
    public var val: Int
    public var left: TreeNode?
    public var right: TreeNode?

    /// Default initializer
    public init() { val = 0; left = nil; right = nil }

    /// Initializer with value
    /// - Parameter val: The value of the node
    public init(_ val: Int) { self.val = val; left = nil; right = nil }

    /// Initializer with value and children
    /// - Parameters:
    ///   - val: The value of the node
    ///   - left: The left child of the node
    ///   - right: The right child of the node
    public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
        self.val = val
        self.left = left
        self.right = right
    }
}

// MARK: - Solution
/**
 A class that provides a solution to the problem of adding a row to a binary tree.
 */
class Solution {
    /**
     Adds a row to a binary tree.

     This function adds a row of nodes with a given value at a given depth in the tree.
     The original nodes at this depth will become the children of the new nodes.

     - Parameters:
       - root: The root of the binary tree.
       - val: The value of the new nodes.
       - depth: The depth at which to add the new row.

     - Returns: The root of the modified tree.

     - Precondition: `depth` > 0

     - Postcondition: The tree has a new row of nodes at the specified depth.

     - Throws: This function throws a fatal error if it cannot add a row at the specified depth.

     - Example:
       ```
       let root = TreeNode(1)
       root.left = TreeNode(2)
       root.right = TreeNode(3)
       let solution = Solution()
       let newRoot = solution.addOneRow(root, 4, 2)
       // The tree now looks like this:
       //     1
       //    / \
       //   4   4
       //  /     \
       // 2       3
       ```
     */
    func addOneRow(_ root: TreeNode?, _ val: Int, _ depth: Int) -> TreeNode? {
        guard depth > 1 else {
            let r = TreeNode(val, root, nil); return r
        }
        var queue = [root!], dp = 2
        while true {
            guard dp == depth else {
                queue = queue.flatMap { [$0.left, $0.right] }.compactMap { $0 }
                dp += 1
                continue
            }
            for item in queue {
                item.left = TreeNode(val, item.left, nil)
                item.right = TreeNode(val, nil, item.right)
            }
            return root
        }
        fatalError()
    }
}
